ts/private/ts_project_worker.js

```diff
@@ -92,41 +124,44 @@ function createFilesystemTree(root, inputs) {
         // NOTE: making a readlink call is more expensive than making a lstat call
         if (stat.isSymbolicLink()) {
             const linkpath = fs.readlinkSync(absolutePath);
-            const absoluteLinkPath = path.isAbsolute(linkpath) ? linkpath : path.resolve(path.dirname(absolutePath, linkpath))
+            const absoluteLinkPath = path.isAbsolute(linkpath) ? linkpath : path.resolve(path.dirname(absolutePath), linkpath)
             return path.relative(root, absoluteLinkPath);
         }
         return p;
     }
 
-    function add(p) {
-        const parts = path.parse(p);        
-        const dirs = parts.dir.split(path.sep).filter(p => p != "");
-
+    function add(p) { 
+        const parts = p.split(path.sep).filter(p => p != "");
         let node = tree;
 
-        for (const [i, part] of dirs.entries()) {
-            if (typeof node[part] != "object") {
-                node[part] = {
-                    [Type]: TYPE.DIR 
-                };
-                notifyWatchers(dirs.slice(0, i), part, TYPE.DIR, EVENT_TYPE.ADDED);
+        for (const [i, part] of parts.entries()) {
+            if (node && node[Type] == TYPE.SYMLINK) {
+                // stop; this is possibly path to a file which points to symlinked treeartifact.
+                // bazel 6 has introduced a weird behavior where it expands treeartifact symlinks when --experimental_undeclared_symlink is turned off.
+                return;
             }
-            node = node[part];
-        }
-
-        const possiblyResolvedSymlinkPath = followSymlinkUsingRealFs(p)
+            const currentP = parts.slice(0, i + 1).join(path.sep)   
+            if (typeof node[part] != "object") {
+                const possiblyResolvedSymlinkPath = followSymlinkUsingRealFs(currentP)
```

Comment:
this is the expensive call that we have to make for each part of path and every input. 

complexity is `number of inputs x length(input.split("/"))`

Reply:
of course, there's tree caching for those inputs that share the same root